tensorflow/lite/kernels/stablehlo_reduce_precision.cc

```diff
@@ -0,0 +1,195 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include <cmath>
+#include <limits>
+#include <type_traits>
+
+#include "Eigen/Core"
+#include "tensorflow/lite/core/c/builtin_op_data.h"
+#include "tensorflow/lite/core/c/common.h"
+#include "tensorflow/lite/kernels/internal/tensor_ctypes.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+
+namespace tflite {
+namespace ops {
+namespace builtin {
+namespace stablehlo_reduce_precision {
+namespace {
+constexpr int kInputTensor = 0;
+constexpr int kOutputTensor = 0;
+
+template <typename DataType>
+struct FloatingPointTraits {
+  static constexpr int kMantissaBits =
+      std::numeric_limits<DataType>::digits - 1;
+  static constexpr int kMaxExponent =
+      std::numeric_limits<DataType>::max_exponent;
+  static constexpr int kMinExponent =
+      std::numeric_limits<DataType>::min_exponent;
+  static constexpr int kExponentBits = (__builtin_clz(static_cast<unsigned int>(
+                                            kMaxExponent - kMinExponent + 1)) ^
+                                        31) +
+                                       1;
+  using UIntType =
+      std::conditional_t<std::is_same_v<DataType, float>, uint32_t, uint16_t>;
+};
+
+template <typename UIntType>
+UIntType ReducePrecision(UIntType int_val, int32_t exponent_bits,
+                         int32_t mantissa_bits, int32_t src_mantissa_bits,
+                         int32_t src_exponent_bits,
+                         UIntType last_mantissa_bit_mask,
+                         UIntType base_rounding_bias, UIntType truncation_mask,
+                         UIntType sign_bit_mask, UIntType exp_bits_mask,
+                         UIntType reduced_max_exponent,
+                         UIntType reduced_min_exponent);
+
+template <typename DataType>
+TfLiteStatus ReducePrecisionOp(const TfLiteTensor* operand,
+                               int32_t exponent_bits, int32_t mantissa_bits,
+                               TfLiteTensor* result) {
+  const int num_elements = NumElements(result);
+  const DataType* input = GetTensorData<DataType>(operand);
+  DataType* output = GetTensorData<DataType>(result);
+  using UIntType = typename FloatingPointTraits<DataType>::UIntType;
+  const int32_t src_mantissa_bits =
+      FloatingPointTraits<DataType>::kMantissaBits;
+  const int32_t src_exponent_bits =
+      FloatingPointTraits<DataType>::kExponentBits;
+
+  UIntType last_mantissa_bit_mask = static_cast<UIntType>(1)
+                                    << (src_mantissa_bits - mantissa_bits);
```

Comment:
If mantissa_bits > src_mantissa_bits, then this could be a shift by a negative number, which would be undefined behavior. We should probably have a default bit mask for the case where src_mantissa_bits is less than mantissa_bits.

Reply:
Okay